# Critical Thinking 5

## Data Scientist Job Market in the U.S.

<br>
Course Code: DS620 <br>
Course Name: Data Visualization <br>
CRN: 15059 <br>
Dr. Mohammad Abdelrahman

Student ID: G200007615 <br>
Student Name: Abdulaziz Alqumayzi<br>
Date: 27/11/2021


### Tasks: 

- Read the data.
- Explore the dataset, make sure there are no missing values, and if there are, they are filled with 0s. 
- Convert data in each column into an appropriate data type.
- Create interactive visualizations that show the location of each geographical column (latitude and longitude). 
- Group the data as you see appropriate.
- Choose and visualize the data using the appropriate chart.

In [67]:
import pandas as pd
import plotly.express as px
import plotly.offline as po
import matplotlib.pyplot as plt
%matplotlib inline
po.init_notebook_mode(connected = True)

### Task: Read the data 

In [68]:
df = pd.read_csv('CV_LatLon_21Jan_12Mar.csv') # import the dataset
df_loc = pd.read_csv('countries.csv') # import location of countries dataset
df

Unnamed: 0           state    country      lat       lon     date  \
0               0               0   Thailand  15.0000  101.0000  1/22/20   
1               1               0      Japan  36.0000  138.0000  1/22/20   
2               2               0  Singapore   1.2833  103.8333  1/22/20   
3               3               0      Nepal  28.1667   84.2500  1/22/20   
4               4               0   Malaysia   2.5000  112.5000  1/22/20   
...           ...             ...        ...      ...       ...      ...   
20701       20701     Oakland, MI         US  42.5922  -83.3362  3/12/20   
20702       20702       Wayne, MI         US  42.2791  -83.3362  3/12/20   
20703       20703  New Castle, DE         US  39.5393  -75.6674  3/12/20   
20704       20704               0       Cuba  22.0000  -80.0000  3/12/20   
20705       20705               0     Guyana   5.0000  -58.7500  3/12/20   

       confirmed  recovered  death  
0              2          0      0  
1              2          0      0  
2              0          0      0  
3              0          0      0  
4              0          0      0  
...          ...        ...    ...  
20701          0          0      0  
20702          0          0      0  
20703          0          0      0  
20704          3          0      0  
20705          1          0      1  

[20706 rows x 9 columns]

### Task: Explore the dataset, make sure there are no missing values, and if there are, they are filled with 0s. 

In [69]:
# high level assessment of the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20706 entries, 0 to 20705
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  20706 non-null  int64  
 1   state       20706 non-null  object 
 2   country     20706 non-null  object 
 3   lat         20706 non-null  float64
 4   lon         20706 non-null  float64
 5   date        20706 non-null  object 
 6   confirmed   20706 non-null  int64  
 7   recovered   20706 non-null  int64  
 8   death       20706 non-null  int64  
dtypes: float64(2), int64(4), object(3)
memory usage: 1.4+ MB


The dataset has 9 columns, 20706 rows and no null values.<br>
The data types are 4 integers, 3 objects and 2 floats. <br><br>
**To-Do:** 
- Create new dataframe contains (country, unified latitude, unified longitude, confirmed, recovered, death) columns

In [70]:
# check statistics of numeric data type
df.describe()

Unnamed: 0           lat           lon     confirmed     recovered  \
count  20706.000000  20706.000000  20706.000000  20706.000000  20706.000000   
mean   10352.500000     34.230418    -39.800212    143.855501     51.206993   
std     5977.451673     16.708021     81.418589   2487.423241   1123.630334   
min        0.000000    -41.454500   -157.858400      0.000000      0.000000   
25%     5176.250000     31.240000    -95.941000      0.000000      0.000000   
50%    10352.500000     38.687450    -77.063950      0.000000      0.000000   
75%    15528.750000     42.409700     17.679100      0.000000      0.000000   
max    20705.000000     64.963100    174.886000  67781.000000  50318.000000   

              death  
count  20706.000000  
mean       4.444847  
std       96.773136  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max     3056.000000

In [71]:
# check how many countries in this dataset
df['country'].nunique()

116

Data transforming and cleaning:

In [72]:
# join the dataset with 
df2 = pd.merge(df,df_loc,how='left', left_on='country', right_on='name')
df2

Unnamed: 0           state  country_x      lat       lon     date  \
0               0               0   Thailand  15.0000  101.0000  1/22/20   
1               1               0      Japan  36.0000  138.0000  1/22/20   
2               2               0  Singapore   1.2833  103.8333  1/22/20   
3               3               0      Nepal  28.1667   84.2500  1/22/20   
4               4               0   Malaysia   2.5000  112.5000  1/22/20   
...           ...             ...        ...      ...       ...      ...   
20701       20701     Oakland, MI         US  42.5922  -83.3362  3/12/20   
20702       20702       Wayne, MI         US  42.2791  -83.3362  3/12/20   
20703       20703  New Castle, DE         US  39.5393  -75.6674  3/12/20   
20704       20704               0       Cuba  22.0000  -80.0000  3/12/20   
20705       20705               0     Guyana   5.0000  -58.7500  3/12/20   

       confirmed  recovered  death country_y   latitude   longitude       name  
0              2          0      0        TH  15.870032  100.992541   Thailand  
1              2          0      0        JP  36.204824  138.252924      Japan  
2              0          0      0        SG   1.352083  103.819836  Singapore  
3              0          0      0        NP  28.394857   84.124008      Nepal  
4              0          0      0        MY   4.210484  101.975766   Malaysia  
...          ...        ...    ...       ...        ...         ...        ...  
20701          0          0      0       NaN        NaN         NaN        NaN  
20702          0          0      0       NaN        NaN         NaN        NaN  
20703          0          0      0       NaN        NaN         NaN        NaN  
20704          3          0      0        CU  21.521757  -77.781167       Cuba  
20705          1          0      1        GY   4.860416  -58.930180     Guyana  

[20706 rows x 13 columns]

The new dataframe needs more processing<b><b>

- Delete Unnamed: 0, state, lat, lon and country_y columns
- Fill missing values in column name
- Change name column to country
- Delete country_x column

In [73]:
# check null values in the name column with the country_x column to fill them up with the right values 
df2.query('name != name')['country_x'].value_counts()

US                  12393
North Macedonia        51
Holy See               51
Korea, South           51
Cruise Ship            51
Czechia                51
Taiwan*                51
Congo (Kinshasa)       51
Cote d'Ivoire          51
Reunion                51
Name: country_x, dtype: int64

In [74]:
pd.set_option('display.max_rows', None) # view all rows
df_loc['name'].value_counts()

Andorra                                         1
Malaysia                                        1
Namibia                                         1
New Caledonia                                   1
Niger                                           1
Norfolk Island                                  1
Nigeria                                         1
Nicaragua                                       1
Netherlands                                     1
Norway                                          1
Nepal                                           1
Nauru                                           1
Niue                                            1
New Zealand                                     1
Oman                                            1
Panama                                          1
Peru                                            1
French Polynesia                                1
Papua New Guinea                                1
Philippines                                     1


The country names are missing:<b>
- United States 
- Macedonia [FYROM]
- Vatican City
- South Korea
- Czech Republic
- Taiwan
- Congo [Republic]
- Côte d'Ivoire
- Réunion
    
The value will be removed because it is not countries: 
- Cruise Ship

In [75]:
# change country names in first dataframe
df["country"].replace({"US": "United States", "North Macedonia": "Macedonia [FYROM]","Holy See":"Vatican City",
                   "Korea, South":"South Korea","Czechia":"Czech Republic","Taiwan*":"Taiwan",
                   "Congo (Kinshasa)":"Congo [Republic]","Cote d'Ivoire":"Côte d'Ivoire","Reunion":"Réunion"}, inplace=True)

# remove Cruise Ship values 
df = df[df["country"] != "Cruise Ship"] 

# should be 115 countries
df['country'].nunique()

115

Repeat the join again

In [76]:
# reset to view the view of rows to accelerate the process
pd.reset_option("^display")
# join the dataset with 
df3 = pd.merge(df,df_loc,how='left', left_on='country', right_on='name')
df3

Unnamed: 0           state      country_x      lat       lon     date  \
0               0               0       Thailand  15.0000  101.0000  1/22/20   
1               1               0          Japan  36.0000  138.0000  1/22/20   
2               2               0      Singapore   1.2833  103.8333  1/22/20   
3               3               0          Nepal  28.1667   84.2500  1/22/20   
4               4               0       Malaysia   2.5000  112.5000  1/22/20   
...           ...             ...            ...      ...       ...      ...   
20650       20701     Oakland, MI  United States  42.5922  -83.3362  3/12/20   
20651       20702       Wayne, MI  United States  42.2791  -83.3362  3/12/20   
20652       20703  New Castle, DE  United States  39.5393  -75.6674  3/12/20   
20653       20704               0           Cuba  22.0000  -80.0000  3/12/20   
20654       20705               0         Guyana   5.0000  -58.7500  3/12/20   

       confirmed  recovered  death country_y   latitude   longitude  \
0              2          0      0        TH  15.870032  100.992541   
1              2          0      0        JP  36.204824  138.252924   
2              0          0      0        SG   1.352083  103.819836   
3              0          0      0        NP  28.394857   84.124008   
4              0          0      0        MY   4.210484  101.975766   
...          ...        ...    ...       ...        ...         ...   
20650          0          0      0        US  37.090240  -95.712891   
20651          0          0      0        US  37.090240  -95.712891   
20652          0          0      0        US  37.090240  -95.712891   
20653          3          0      0        CU  21.521757  -77.781167   
20654          1          0      1        GY   4.860416  -58.930180   

                name  
0           Thailand  
1              Japan  
2          Singapore  
3              Nepal  
4           Malaysia  
...              ...  
20650  United States  
20651  United States  
20652  United States  
20653           Cuba  
20654         Guyana  

[20655 rows x 13 columns]

In [77]:
# rename the column name to country
df3.rename(columns={'name':'country'}, inplace=True)

In [78]:
# check null values in the name column with the country_x column 
df3.query('country != country')['country_x'].value_counts() # should not return anything

Series([], Name: country_x, dtype: int64)

### Task: Convert data in each column into an appropriate data type. (Not Applicable)
The final dataset (df_final) contains the right data types 

### Task: Group the data as you see appropriate.

In [79]:
df_final = df3.groupby(['country','latitude','longitude'])['confirmed','recovered','death'].sum().reset_index()
df_final

C:\Users\amg11\AppData\Local\Temp/ipykernel_21784/354187595.py:1: FutureWarning:

Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.



country   latitude   longitude  confirmed  recovered  death
0             Afghanistan  33.939110   67.709953         40          0      0
1                 Albania  41.153332   20.168331         47          0      2
2                 Algeria  28.033886    1.659626        175          8      1
3                 Andorra  42.546245    1.601554         11          1      0
4               Argentina -38.416097  -63.616672         92          0      5
..                    ...        ...         ...        ...        ...    ...
110  United Arab Emirates  23.424076   53.847818        802        161      0
111        United Kingdom  55.378051   -3.435973       2842        275     34
112         United States  37.090240  -95.712891       6984        179    216
113          Vatican City  41.902916   12.453389          7          0      0
114               Vietnam  14.058324  108.277199        698        406      0

[115 rows x 6 columns]

### Task: Create interactive visualizations that show the location of each geographical column (latitude and longitude).  

In [80]:
map_fig_1 = px.scatter_geo(df_final, lat='latitude',
                        lon='longitude',
                        color='confirmed',
                        projection='orthographic',
                        title='Confirmd Covid-19 Map',      
                        hover_name='country',
                        hover_data=['confirmed','recovered','death'])
map_fig_1.update_traces(marker=dict(size=10))
map_fig_1.show()

In [81]:
map_fig_1 = px.scatter_geo(df_final, lat='latitude',
                        lon='longitude',
                        color='recovered',
                        projection='orthographic',
                        title='Recovered Covid-19 Map',      
                        hover_name='country',
                        hover_data=['confirmed','recovered','death'])
map_fig_1.update_traces(marker=dict(size=10))
map_fig_1.show()

In [82]:
map_fig_1 = px.scatter_geo(df_final, lat='latitude',
                        lon='longitude',
                        color='death',
                        projection='orthographic',
                        title='Death Covid-19 Map',      
                        hover_name='country',
                        hover_data=['confirmed','recovered','death'])
map_fig_1.update_traces(marker=dict(size=10))
map_fig_1.show()

### Task: Choose and visualize the data using the appropriate chart.

In [83]:
# confirmed of covid-19 per countries
fig_1 = px.bar(df_final, y='confirmed', x='country', text='confirmed')
fig_1.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig_1.show()

In [84]:
# recovered of covid-19 per countries 
fig_2 = px.bar(df_final, y='recovered', x='country', text='recovered')
fig_2.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig_2.show()

In [85]:
# death of covid-19 per countries 
fig_3 = px.bar(df_final, y='death', x='country', text='death')
fig_3.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig_3.show()

#### References
- Belorkar, A., Guntuku, S. C., Hora, S., & Kumar, A. (2020). Interactive data visualization with python: Present your data as an effective and compelling story. Packt Publishing. 
- Maliarenko, M. (2020, March 13). Coronavirus Lat/Lon Dataset. Kaggle. Retrieved November 27, 2021, from https://www.kaggle.com/grebublin/coronavirus-latlon-dataset. 
- Google&nbsp;|&nbsp; dataset publishing language &nbsp;. (2021). Countries.csv . google developers. Retrieved November 27, 2021, from https://developers.google.com/public-data/docs/canonical/countries_csv. 
- So, B. (2020, February 6). 2019 coronavirus dataset (January - February 2020). Kaggle. Retrieved November 27, 2021, from https://www.kaggle.com/brendaso/2019-coronavirus-dataset-01212020-01262020. 

